fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMerger.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * IndexingMerger.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.FDBError;
+import com.apple.foundationdb.FDBException;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+/**
+ * Implement an index merge.
+ * This can be either used for an explicit index merge call or during online indexing. In the second case,
+ * keeping throttle values between calls has an overall impact on performance.
+ */
+@ParametersAreNonnullByDefault
+public class IndexingMerger {
+    private static final Logger LOGGER = LoggerFactory.getLogger(IndexingMerger.class);
+    private final Index index;
+    private long mergesLimit;
+    private int mergeSuccesses = 0;
+    private final IndexingCommon common;
+
+    // These error codes represent a list of errors that can occur if there is too much work to be done
+    // in a single transaction.
+    private static final Set<Integer> lessenWorkCodes = new HashSet<>(Arrays.asList(
```

Comment:
It could be easily shared, but there is a chance that at some point we'll want different error sets for the two type of retries. 
Saying that, I'll simply share it and will split again if and when that day will ever come.